#6722 initialize proxies with their to-many fields already initialized to a `PersistentCollection` instance

Ocramius opened the pull request:
Implements #6722

Ocramius committed:
Removing functional test cases covering `EntityManager#merge()` semantics

Ocramius committed:
Removing tests around `EntityManager#merge()` and similar operations

Ocramius committed:
Removing `EntityManager#merge()` logic

Ocramius committed:
Removing merge operations from the UnitOfWork

Ocramius committed:
Adding TODO: serialization of entities is something the ORM shouldn't deal with (directly) anymore

Ocramius committed:
Adding @TODO about batch-processing and `detach()` operations: `clear()` may deal with it directly.

Ocramius committed:
Note about sections to be removed

Ocramius committed:
Removing tests around the `detach` functionality

Ocramius committed:
Removing `EntityManager#detach()`

Ocramius committed:
Disabling test around the `EntityManager#clear($entityName)` functionality (improvement needed)

Ocramius committed:
Removing `clear($entityName)` functionality (for now)

guilhermeblanco committed:
Removed isIdentifierComposite flag from ClassMetadata

guilhermeblanco committed:
Quick rename

guilhermeblanco committed:
Work on ClassMetadataFactory. Updated ProxyFactory reducing dependencies

guilhermeblanco committed:
Refactored Proxy support

guilhermeblanco committed:
Quick fixes

Ocramius committed:
#6118 removing DDC-720 related tests

Ocramius committed:
#6118 removing DDC-720 related tests for multiple flushed entities

Ocramius committed:
#6118 remove reference to single entity flush operations

Ocramius committed:
#6118 remove `flush`\($entity\)` signature

Ocramius committed:
#6118 removing `flush\($entity\)` usage from the test suite

Ocramius committed:
#6118 removing `UnitOfWork#commit\($entity\)` signature

guilhermeblanco committed:
Fix bugs related to rebase

guilhermeblanco committed:
Fixed tetss after rebase

guilhermeblanco committed:
Removed Builders

Ocramius committed:
#6118 upgrade notes mentioning the BC break in `EntityManager#flush()`

Ocramius committed:
#1577 remove newly introduced tests around merge/detach semantics, which were dropped

Ocramius committed:
#1577 adding upgrade notes about the BC break caused by merge() and detach() semantics removal

Ocramius committed:
#1577 re-adding `merge()` and `detach()`, since a release of doctrine/common is required to completely remove those methods from the `ObjectManager` interface first

Ocramius committed:
#1577 re-wording 'working with sessions' section, so that users are discouraged from storing entities in sessions

Ocramius committed:
#1577 removing `detach()` calls references from the batch processing docs, adding links to existing batch processing utilities

Ocramius committed:
#1577 removing <cascade-merge/> and <cascade-detach/> from the XSD

Ocramius committed:
#1577 removing cascade-merge and cascade-detach from mapping examples

Ocramius committed:
#1577 removing cascade-merge and cascade-detach from mapping examples

Ocramius committed:
#1577 removing references to merge/detach cascade operations

Ocramius committed:
#1577 removing documentation about merge/detach methods, replaced by clear() documentation only

Ocramius committed:
#1577 removing all references to `merge` operations in the docs

Ocramius committed:
#1577 `merge` and `detach` are no longer accepted cascade mappings

Ocramius committed:
#1577 removing `merge` and `detach` related mapping operations in cascades in tests

Ocramius committed:
Corrected `XmlExporter` order of exported cascades

Majkl578 committed:
Fix type errors

Majkl578 committed:
Enable strict types

Ocramius committed:
#1577 re-enabling `clear($entityName)` tests

guilhermeblanco committed:
Quick fix

guilhermeblanco committed:
Added generation strategy for class metadata

guilhermeblanco committed:
Removed Instantiator from ClassMetadata and concentrate into UnitOfWork

guilhermeblanco committed:
Removed sequence methods from class metadata

Majkl578 committed:
OrmTestCase: Always use full AnnotationReader, drop legacy versions

Majkl578 committed:
Populate FieldMetadata ID generator

Majkl578 committed:
Move identifier generation to FieldMetadata

Majkl578 committed:
WIP

guilhermeblanco committed:
Revert "[3.0] Moved IdentifierGenerator to Property."

guilhermeblanco committed:
Moved IdentifierFlattener temporarily to EntityManager and removed one todo

guilhermeblanco committed:
Optimizations around EntityPersisters

Majkl578 committed:
Added ValueGeneratorMetadata, implemented ValueGeneratorPlan

guilhermeblanco committed:
Removed enqueueing for insert support from EntityPersister. There was no reason for that since concurrent inserts NEVER guarantee incremental order on a single INSERT statement. Any attempt to reliably implement is not accurate and thus I am removing any possibility of this ever becoming a reality.

guilhermeblanco committed:
Return array of identifier field names instead of hardcoded identifier

guilhermeblanco committed:
Ensure we are using getClassName() which will be later available in ComponentMetadata

guilhermeblanco committed:
Removed direct access to ClassMetadata->name

guilhermeblanco committed:
Removed direct access to ClassMetadata->name

guilhermeblanco committed:
Removed EntityListenerBuilder

guilhermeblanco committed:
Consume custom repository class through methods only.

lcobucci committed:
Replace $this->assertXXX with self::assertXXX

Since PHPUnit test methods are all static

guilhermeblanco committed:
Simplified NamedQuery support

guilhermeblanco committed:
Implemented runtime resolution of __CLASS__ calls to SqlResultSetMapping, NativeNamedQuery and NamedQuery. Dropped isSelfClass from mappings

guilhermeblanco committed:
Simplified logic and stored information around named native query.

guilhermeblanco committed:
Simplified RSM building for result class

guilhermeblanco committed:
Moved MappingDriver to ORM to allow changes in its API

guilhermeblanco committed:
Eliminated the concept of ClassMetadataInterface temporarily until refactoring is complete. Replaced ObjectManagerAware with EntityManagerAware sinc eit was bound to the Common\Persistence namespace with hard dependencies. The same happened to PersistentObject.

guilhermeblanco committed:
Decoupled Generator and GeneratorStrategy

guilhermeblanco committed:
Moved computation of FCQN resolution to Metadata building step.

guilhermeblanco committed:
Moved getIdentifierValues to EntityPersister

guilhermeblanco committed:
Moved ClassMetadata::assignIdentifier to EntityPersister::setIdentifier. Renamed EntityPersister::getIdentifierValues to EntityPersister::getIdentifier

Majkl578 committed:
Drop AssignedGenerator (obsoleted by ValueGenerationPlan)

guilhermeblanco committed:
Fixed issue with associatied entities implementing __toString()

guilhermeblanco committed:
Added information of action items for later.

guilhermeblanco committed:
Added concept of SecondPass to ClassMetadataFactory

guilhermeblanco committed:
More optimizations and preparation for ClassMetadataFactory replacement. Integration with ComponentMetadata started.

Commit:
added the final keyword EntityManager to indicate that extending EntityManager is not a valid extension point.

Majkl578 committed:
Typehint EntityManagerInterface instead of EntityManager where possible

Majkl578 committed:
Typehint EntityManagerInterface instead of EntityManager where possible

Commit:
added the test and updated the entitymanager

mikeSimonson committed:
removing copyright headers in every files

mikeSimonson committed:
Remove another copyright header

mikeSimonson committed:
Add missing strict type declaration

Commit:
fixed the broken tests and updated the UPGRADE guide to notify about BC

Commit:
change the em property to fix the tests and UPGRADE file as per the PR comments.

guilhermeblanco committed:
Some quick typehints

Commit:
Reverted the public method on construct to protected. Updated the tests.

Commit:
updated the heading for upgrade doc

guilhermeblanco committed:
ClassMetadataFactory changed to bring ClassMetadata instantiation closer to MappingDriver loading call. Changed how ReflectionService is called for ClassMetadata initialization, simplifying calls.

guilhermeblanco committed:
More simplifications in ClassMetadataFactory.

guilhermeblanco committed:
Removed PHP and Static PHP mapping drivers, since they will be irrelevant with the new ClassMetadataFactory

guilhermeblanco committed:
Fixed broken tests

guilhermeblanco committed:
Moved class metadata binding to proper methods

Ocramius committed:
Removing `count()` operations from the `AbstractQuery` implementations

Ocramius committed:
Removing useless `count()` operations from the `PersistentCollection` internals

Ocramius committed:
Removing useless counting around query resultset handling

Ocramius committed:
Removing counting operations around parameter handling in `QueryBuilder`

Ocramius committed:
Removing counting operations in the `UnitOfWork`

Ocramius committed:
Avoiding counting aliases when determining whether a query is simple or not

Ocramius committed:
Avoid repeated counting of root aliases in `ArrayHydrator`

Ocramius committed:
Avoiding counting scalars at every iteration - simply checking once is enough

Ocramius committed:
Removing counting of generated id fields when checking for composite identifiers

Ocramius committed:
Quicker check on whether an identifier is composite - removes counting

Ocramius committed:
Removing useless counting of the discriminator map during metadata loading

Ocramius committed:
Fixing check of inheritance on a mapped superclass

Ocramius committed:
Removing useless counting on whether subclasses are present

Ocramius committed:
Removed counting of local variable when generating filter clauses for many to many collections

Ocramius committed:
Removing counting from `getSelectConditionStatementSQL`

Ocramius committed:
Removing some useless counting operations from the DQL parser

Ocramius committed:
Removing counting in an iteration when generating aliases

Ocramius committed:
Removed counting of sql parts when deciding whether to wrap parts in parentheses

Ocramius committed:
Removed typical `if (count(expr)) {` issue from sql walker

Ocramius committed:
Avoiding fallback function lookup when counting expression parts in a generic query builder expression

Ocramius committed:
Avoiding fallback function lookup when counting expression parts in an `OrderBy` query builder expression

Ocramius committed:
Reverting dangerous index lookup operations - resultset may contain gaps in the results

Ocramius committed:
Replacing possible O(n^2) operation from `PersistentCollection` diffing operations

Ocramius committed:
Removing useless checking for never-used parameter, inlining merge operation

Ocramius committed:
Removing useless variable

Ocramius committed:
Removing useless post-array-population association hydration via `array_walk()`

Ocramius committed:
Removing useless `array_key_exists()` calls, using `??` operator instead

Ocramius committed:
Removing `is_array` check on a type that can only have two possible states - using `instanceof` instead

guilhermeblanco committed:
TableMetadata should only exist if ClassMetadata has a table. Started some concept work around Mapping Binders.

Ocramius committed:
Adding `ocramius/package-versions` as a replacement for the internal ORM `Version` class

Ocramius committed:
Dropping references to the ORM `Version` class - using `PackageVersions` instead

Ocramius committed:
Removing build references to the `Version` class, which is finally gone-gone-gone

Ocramius committed:
#6653 adding note to `UPGRADE.md` about the removal of `Doctrine\ORM\Version`

Ocramius committed:
#1577 removing references to `detach` and `merge` cascade operations in newer metadata driver implementations

Ocramius committed:
#1577 removed `detach` reference from the docs, as per @lcobucci's review

Ref: https://github.com/doctrine/doctrine2/pull/1577#discussion_r124127083

Ocramius committed:
#1577 rephrasing reference to `detach` that is now `clear()` instead

Ref: https://github.com/doctrine/doctrine2/pull/1577#discussion_r124126422 - @lcobucci's review

Ocramius committed:
#1577 clarifying session user fetching as per @Majkl578's review

Ref: https://github.com/doctrine/doctrine2/pull/1577#discussion_r124162444

Ocramius committed:
#6653 removing unused variable as per @alcaeus' review

Ref: https://github.com/doctrine/doctrine2/pull/6653#discussion_r135503849

Majkl578 committed:
Use Generators instead of AppendIterator for properties iteration

Ocramius committed:
Stubbing out a retry on the s/`Proxy`/`GhostObjectInterface` replacement patch

Ocramius committed:
Ensuring that the `BasicFunctionalTest` relies on the `GhostObjectInterface` API

Ocramius committed:
Prevent setting any values in un-initialized proxies

Ocramius committed:
Proxy identifier must be set when the proxy is instantiated

Ocramius committed:
Correcting `ProxyFactoryTests`, which were opinionated around past proxy definition

Ocramius committed:
Fixing faulty test that was expecting a different entity to bbe hydrated during proxy initialization

Ocramius committed:
Correcting reference to non-existing method

Ocramius committed:
ORM proxy names may be looked up in the class metadata factory - the parser should not manually do `class_exists` checks unless strictly required

Ocramius committed:
Hardening `ProxiesLikeEntitiesTest` by adding some precondition checks

Ocramius committed:
Cascade operations do not apply to non-initialized proxies: no associations are set there

Ocramius committed:
Transient fields should be skipped by proxy initialization semantics

Ocramius committed:
Fixed reference to non-existing method

Ocramius committed:
Cleaning up `ReferenceProxyTest`, which no longer reflects current invariants

Ocramius committed:
Correcting `DDC1193Test`, which was using the old `Proxy` API

Ocramius committed:
Corrected `DDC1228Test`, which was quite broken and relying on the old `Proxy` API

Ocramius committed:
Correcting `DDC1690Test`: even if a proxy is not yet loaded, it must have notify property listeners to function correctly

Ocramius committed:
The `DefaultQueryCache` still contained `instanceof Proxy` checks - replaced with `GhostObjectInterface` checks

Ocramius committed:
Correcting `DDC2231Test`, since proxies no longer initialize if no state is mutated

Ocramius committed:
Hardening DDC3223Test with additional pre- and post-conditions

Ocramius committed:
Removing DDC-1238 related code. Issue seems to be invalid, as cloning a proxy should NOT modify its state, but just make it non-managed

Ocramius committed:
Documenting DDC-1022 tests to be removed, as proxy behavior changed around `__clone` and `__wakeup` semantics

Ocramius committed:
Correcting DDC-1238 to remove non-sensical semantics that cleared proxy data after detach operations

Ocramius committed:
Correcting metadata factory tests - fetching a class by name does indeed fetch the class even if prefixed

Ocramius committed:
Adding `ocramius/proxy-manager` dependency

Ocramius committed:
`SecondLevelCacheManyToOneTest` should not cause more queries for fetching associated identifiers (which are already available)

Ocramius committed:
Adding stub implementation of logic that converts a flat identifier into a deep identifier

Ocramius committed:
Explicit BC break: we no longer call `__wakeup` on proxies

Ocramius committed:
Correcting `ReferenceProxyTest`: proxy identifiers are no longer unset while cloning, but they are not considered managed

Ocramius committed:
Stop registering cloned proxies (real instance being managed) as managed

Ocramius committed:
The `NotifyPropertyChanged` API should only be used once we decided to overwrite the entity data, in which case the `UnitOfWork` is indeed tracking entity state

Ocramius committed:
Replace `ClassUtils` with a temporary, local, deprecated shim that knows about `ProxyManager` proxy class names

Ocramius committed:
Beginning work on the `Configuration` overhaul - we want to manage a `ProxyManager\Configuration` instance in the ORM config

Ocramius committed:
Setting a proxy path should update the `ProxyManager\Configuration` instance too

Ocramius committed:
Extracting temporary dir generation to a private method

Ocramius committed:
Changing auto-generation flag should also change the `ProxyManager\Configuration` referenced generator strategy

Ocramius committed:
Verifying that the `Doctrine\ORM\Configuration` builds separate and independent ghost object factory instances

Ocramius committed:
Adding `Doctrine\ORM\Configuration` API as per test specification

Ocramius committed:
`setProxyNamespace()` should also update the wrapped `ProxyManager\Configuration` instance

Ocramius committed:
`SetupTest` was using invalid proxy paths

Ocramius committed:
Rewriting `StaticProxyFactory` so its constructor just receives a `GhostObjectFactory` as constructor argument, and that's it

Ocramius committed:
Using the `LazyLoadingGhostFactory` to trigger proxy generation

Ocramius committed:
Improving performance of the `StaticProxyFactory` by caching all internal components that do not require repeated fetch/build steps

Ocramius committed:
Removing all unused proxy-related components

Ocramius committed:
Replacing `ClassUtils` shim usage with a new `StaticClassNameConverter` with less features

Ocramius committed:
Basic tests for the `StaticClassNameConverterTest`

Ocramius committed:
Verifying all of the `StaticClassNameConverterTest` API

Ocramius committed:
Removed `ClassUtils` shim: class has been internalized

Ocramius committed:
Dead code removal

Ocramius committed:
Removing unused import

Ocramius committed:
Removing solved `TODO` in the `Parser` changes

Ocramius committed:
Removing solved TODOs and improved documentation around why we work with an un-managed proxy anyway

Ocramius committed:
Removing duplicate `TODO` that was moved to a private method documentation

Ocramius committed:
Simplifying `generateProxyClasses` code

Ocramius committed:
Simplified spec for an identifier normalizer

Ocramius committed:
Completing specification for nested reference identifier normalization

Ocramius committed:
Implemented `NormalizeIdentifier` as per specification

Also made sure the `EntityManager` does not ruin identifiers when `getReference` is being called

Ocramius committed:
Using `NormalizeIdentifier` inside the `UnitOfWork` instead of relying on an additional internal method

Ocramius committed:
Minimal `NormalizeIdentifier` documentation

Ocramius committed:
Inlining variables that are only used once

Ocramius committed:
Marking utility classes as internal

Ocramius committed:
Extracting `NormalizeIdentifierTest` assets to a `Model` namespace

Ocramius committed:
Verifying that identifiers are preserved when using `EntityManager#getReference()`

Ocramius committed:
Verifying that identifier order being swapped does not cause identifier references to be lost

Ocramius committed:
Removing fixed TODO about identifier handling

Ocramius committed:
Verifying that ghost object friend objects are not initialized when not reading lazy state

Ocramius committed:
Verifying friend object "Comparable" behavior when using proxies that access each other's private state

Ocramius committed:
Using `AUTOGENERATE_EVAL` during test suite runs

Ocramius committed:
Adapting proxy configuration documentation to the new changes in the proxy handling in ORM v3

Ocramius committed:
Removing redundant/conflicting proxy documentation, correcting proxy performance/autoloading docs

Ocramius committed:
`func_get_args()` now works within proxies: YAY!

Ocramius committed:
Updating documentation about proxy serialization

Ocramius committed:
s/lazy-load/lazy-loading

Ocramius committed:
Removing FAQ entry for item that no longer holds true about `public` property lazy-loading semantics

Ocramius committed:
Removing documented quirks about proxy lazy-loading semantics combined with reflection: now always holding true

Ocramius committed:
Removing reference to previous proxy model

Ocramius committed:
Replacing generated proxy example with a stub resembling current state

Ocramius committed:
Cleaning documentation around proxy behavior

Ocramius committed:
Removing erroneous documentation about `final` methods, `c__lone` and `__wakeup` details that no longer hold true

Ocramius committed:
Marking the `StaticProxyFactory` as internal, fixing type hint documentation

Ocramius committed:
Documenting introduced BC breaks in the proxy refactoring

Ocramius committed:
Clarifying that also reflection now triggers lazy-loading

Ocramius committed:
Cleaning up performance tests before attempting a fix of the `ProxyInitializationTimeBench`

Ocramius committed:
Ignoring `phpbench.phar` and the associated public key

Ocramius committed:
Removing `NonLoadingPersister` specifics that are not to be used in benchmarks

Ocramius committed:
Proxies should be initialized with all their to-many associations pre-filled with collections

Ocramius committed:
Cleaning up last tests that were still using real filesystem for proxies

Ocramius committed:
Specification for the feature: accessing to-many associations in proxies should not lazy-load the proxy itself

Ocramius committed:
Proxies now skip lazy-loading for to-many associations

Ocramius commented on lib/Doctrine/ORM/Proxy/Factory/StaticProxyFactory.php:
```diff
@@ -127,6 +138,15 @@ public function getProxy(string $className, array $identifier) : GhostObjectInte
+        foreach ($toManyFields as $toMany) {
+            $collection = $toMany->wrap($proxyInstance, null, $this->entityManager);
+
+            $collection->setInitialized(false); // @TODO not the right approach, IMO
```
Seems correct, although I'd prefer the `$toMany` `Property` to do this for me :-\

stof replied:
shouldn't this be done in `->wrap()` instead ? It looks weird to me that wrapping `null` gives you an initialized PersistentCollection.

Ocramius replied:
Yeah, TODO in fact. `wrap` was not working as expected either.

Ocramius commented on lib/Doctrine/ORM/Proxy/Factory/StaticProxyFactory.php:
```diff
@@ -186,6 +207,44 @@ private function transientFieldsFqns(ClassMetadata $metadata) : array
+    {
+        $toMany = [];
+
+        foreach ($metadata->getDeclaredPropertiesIterator() as $property) {
```
Note: not using `array_map()` because this is an iterator

guilhermeblanco replied:
There're several pro/cons on Iterator for properties. Keep in mind that `getDeclaredProperties` is different from `getProperties` as the latter also covers hierarchy.

I have yet to decide if we'll make it an iterator forever or no.

Ocramius replied:
Ideally, I'd like to have pre-filtered iterators for things like identifiers, association types, etc 😁

guilhermeblanco commented on lib/Doctrine/ORM/Proxy/Factory/StaticProxyFactory.php:
```diff
@@ -112,6 +121,8 @@ public function getProxy(string $className, array $identifier) : GhostObjectInte
                 ->entityManager
                 ->getUnitOfWork()
                 ->getEntityPersister($metadata->getClassName());
+        $toManyFields = $this->cachedToManyFields[$className]
```
Blank line yo

Ocramius replied:
Same group of assignments, hence no added newline

guilhermeblanco commented on lib/Doctrine/ORM/Proxy/Factory/StaticProxyFactory.php:
```diff
@@ -186,6 +207,44 @@ private function transientFieldsFqns(ClassMetadata $metadata) : array
+                $property
+                    ->getDeclaringClass()
+                    ->getReflectionClass()
+                    ->getProperty($name) // @TODO possible NPR. This should never be null, why is it allowed to be?
```
and it's not. unless you got a bug. =\

Ocramius replied:
Yes, the problem is that `getProperty()` can lead to a NPE here. Would probably need an API that gives me a reflection property or throws

Ocramius replied:
Indeed, accessor that throws and has a non-null return type is needed.

Majkl578 replied:
Get method should throw imho, unlike find.

Ocramius committed:
Adding missing trailing comma as per @jaapio's review

Ref: https://github.com/doctrine/doctrine2/pull/6719#discussion_r140642661

Ocramius committed:
BC Break: Removing `getProxyDir`, `getProxyNamespace` and `getAutoGenerateProxyClasses` methods

Also cleans up the `Doctrine\ORM\Configuration` class, which used weird array access to save some memory (no longer true since PHP 5.4+)

Ocramius committed:
Documented removal of `getProxyDir`, `getProxyNamespace` and `getAutoGenerateProxyClasses` methods

Ocramius committed:
Making sure the `ProxyFactory#createProxy()` is always given a `ClassMetadata` instance

Ref: https://github.com/doctrine/doctrine2/pull/6719#discussion_r140646281

Ocramius committed:
Static analysis fixes as reported by scrutinizer-ci

Ocramius commented:
This is to be checked: not sure if collection updates are picked up for un-initialized proxies.

SenseException committed:
Remove copy method from EntityManager

carusogabriel committed:
Refactoring more tests

guilhermeblanco committed:
Fixes after rebase

Ocramius committed:
Verifying that the entity in DDC2231Test is actually `EntityManagerAware`

Ocramius committed:
Forgot to set the `EntityManager` instance also when dealing with `EntityManagerAware` objects (was only using `ObjectManagerAware`)

Ocramius committed:
Hardening DDC2231 test - both `ObjectManagerAware` and `EntityManagerAware` should work

Ocramius committed:
Ensuring that `ObjectManagerAware` and `EntityManagerAware` are equally handled

Ocramius committed:
Ensuring that the `ObjectManagerAware` and `EntityManagerAware` objects are initialized even if they are un-initialized proxies

Ocramius committed:
#6867 as per discussion with @guilhermeblanco, dropping `ObjectManagerAware` support, as the `ClassMetadata` API radically changed

Ocramius committed:
Merge pull request #6867 from doctrine/fix/entity-manager-aware-test-fixes

Ensure `ObjectManagerAware` and `EntityManagerAware` are correctly handled by UoW load/refresh

Majkl578 committed:
Merge pull request #6854 from carusogabriel/refactoring-tests

Refactoring tests

Majkl578 committed:
Merge pull request #6794 from SenseException/remove-copy-method

Remove copy() method from EntityManager

Majkl578 commented:
Needs a rebase.

Majkl578 committed:
Remove code generation tools & commands

carusogabriel committed:
Use Null Coalesce Operator

Ocramius committed:
Merge pull request #6870 from Majkl578/dev/remove-code-generators

Remove code generation tools & commands

Majkl578 committed:
Also run Scrutinizer under PHP 7.2

Majkl578 committed:
Merge pull request #6874 from carusogabriel/null-coalesce-operator

Use Null Coalesce Operator

Majkl578 committed:
Merge pull request #6877 from Majkl578/scrutinizer-7.2

Also run Scrutinizer under PHP 7.2

guilhermeblanco committed:
Ephemeral UnitOfWork

Ocramius committed:
#6722 simplified test models for a uni-directional `ManyToMany` association

Ocramius committed:
#6722 ensuring that the `EntityManagerMock` and `UnitOfWorkMock` never reference a real `UnitOfWork`, but only each other

Ocramius committed:
#6722 verifying that uninitialized proxies are also checked for to-many association field changes (collections)

Ocramius committed:
#6722 ensuring that the `UnitOfWork` crawls over `ToMany` association fields of proxies whilst not attempting to fetch lazy fields

Ocramius commented on lib/Doctrine/ORM/Proxy/Factory/StaticProxyFactory.php:
```diff
@@ -175,6 +196,44 @@ private function transientFieldsFqns(ClassMetadata $metadata) : array
+        $transientFieldsFqns = [];
+
+        foreach ($metadata->getDeclaredPropertiesIterator() as $name => $property) {
+            if (! $property instanceof ToManyAssociationMetadata) {
```
I don't like this kind of filtering of an iterator - do want a dedicated metadata API endpoint

Majkl578 replied:
Thinking about how to avoid ton of getDeclared*() methods, quick idea: what would you say to PropertyFilterIterator + binary flags to pick your own set + factory methods (forAssociations())?

Ocramius replied:
Bad idea: lots of untyped methods then. Having multiple dedicated accessors for metadata isn't a problem IMO.

Ocramius commented on lib/Doctrine/ORM/UnitOfWork.php:
```diff
@@ -745,6 +774,73 @@ public function computeChangeSets()
         }
     }
 
+    private function computeToManyAssociationChanges(ToManyAssociationMetadata $association, $value)
```
This code is mostly duplicated/extracted from `computeAssociationChanges`. The code duplication can be massively reduced if we use dedicated iterators over property types.

Ocramius commented on lib/Doctrine/ORM/UnitOfWork.php:
```diff
@@ -673,11 +687,31 @@ public function computeChangeSet(ClassMetadata $class, $entity)
+                }
+
+                $this->computeToManyAssociationChanges($property, $value);
+            } else {
```
Didn't find a cleaner way for now. Also here, iterating over dedicated accessors would be better.

ostrolucky replied:
`elseif ($property instanceof AssociationMetadata && !$isUnInitializedProxy)` :upside_down_face:

Ocramius replied:
Aye, but the `if` blocks still need de-duplication anyway

Ocramius commented:
Argh, I'll need to rebase this

Pierstoval commented:
Yep, you do 😛

derrabus commented:
I'm closing this PR because it hasn't seen progress for several years. Please open a new one if you want to pick up the work one day.
